Fix #1496: Allow partial paths in jsDependencies.

sjrd opened the pull request:
Partial paths written in `jsDependencies` (including just file names) are shorthands for the full relative paths wrt. the classpath entries.

Different partial paths resolving to the same full relative paths are considered equivalent for dependency resolution and ordering.

If a partial path does not resolve to a unique full relative path on the classpath, a `JSLibResolveException` is thrown.

sjrd commented:
This is the alternative fix to #1496 discussed in the previous PR #1509.
Review by @gzm0 
/cc @fxmouthuy

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1450/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1450/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1856/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1856/</a>
Test PASSed.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/jsdep/JSDependencyManifest.scala:
```diff
@@ -90,8 +91,8 @@ object JSDependencyManifest {
       y.commonJSName.isDefined &&
-      (x.resourceName == y.resourceName ^
-       x.commonJSName == y.commonJSName)
+      (x.relPath == y.relPath ^
```
This needs proper prefix matching. Otherwise this will fail:

```
"jasmine.js" commonJSName "jasmine"
"foo/bar/jasmine.js" commonJSName "jasmine"
```

Even if they refer to the same file.

gzm0 replied:
Ok. I realized that this actually works as it should.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/classpath/PartialClasspath.scala:
```diff
@@ -56,23 +59,78 @@ final class PartialClasspath(
 
-    for (info <- includeList)
-      yield new ResolvedJSDependency(availableLibs(info.resourceName), info)
+    resolvedLibs.result()
```
please add an else clause

sjrd replied:
eh?

gzm0 replied:
I think we should prefer
```
if (...)
  throw
else
  result
```

to

```
if (...)
  throw

result
```

sjrd replied:
Ah that's what you meant. Well I don't have any strong feeling either way in this case, in the sense that the result of this method can be thought of as being either a map or throwing an exception. But I'm pretty sure that so far, the throws have not been accompanied by an `else`.

I'll change it in the morning.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/jsdep/FlatJSDependency.scala:
```diff
@@ -7,11 +7,15 @@ import org.scalajs.core.ir.Trees.isValidIdentifier
     val origin: Origin,
-    val resourceName: String,
+    @deprecatedName('resourceName)
+    val relPath: String,
```
We should somehow document that this `relPath` is not the same than the one in `JSDependency`.

gzm0 replied:
Oh... actually it is called `resourceName` in `JSDependency`. Could you just add a comment there that it may also be a partial relative path?

sjrd replied:
It cannot. It must be a full relative path.

sjrd replied:
I have made this more explicit in the ScalaDoc.

gzm0 commented:
Could you add a test project in `sbt-plugin-test` that does a couple of funky resolves? (Also with local JS files and overrides if possible).

gzm0 commented:
That's all.

sjrd commented:
> Could you add a test project in sbt-plugin-test that does a couple of funky resolves? (Also with local JS files and overrides if possible).

OK, I'll do that.

sjrd commented:
Updated.

I addressed everything except the "else clause", because I don't understand what you're talking about ^^

Once this is in, I'll publish 0.6.1.

scala-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1453/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1453/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1862/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1862/</a>
Test FAILed.

scala-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1454/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1454/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1863/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1863/</a>
Test FAILed.

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1452/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1452/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1861/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1861/</a>
Test PASSed.

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1455/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1455/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1864/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1864/</a>
Test PASSed.

gzm0 commented:
LGTM, once the `else` is changed.
There is one more problem we need to solve, but this can wait for 0.6.2:

If a library (as in Scala.js library) I use depends on, say, `jQuery.js`, and I need a webjar (or library) that bundles `jQuery.js` as well, there is no way I can resolve this conflict (before that, you could use filter for everything. But now it comes after the first resolution step).

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1456/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1456/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1866/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1866/</a>
Test PASSed.
